### ISSUE TITLE
Style/FormatStringToken false positive

### ISSUE THREAD

**Issue by pegasd** [bbatsov/rubocop#5245]
I came across the following false positive while doing rspec-puppet testing on an sshd module.
Sometimes contents of a file can include valid string tokens that look like unannotated Ruby tokens.
These are usually wrapped inside a RegEx for testing.

For example, `sshd_config` can include `%u` which is a valid username reference.

## Expected behavior

The following code shouldn't be flagged as offensive:

```
it {
  is_expected.to contain_file('/etc/ssh/sshd_config')
    .with_content(%r{^    ChrootDirectory /var/chroot/%u$})
}
```

## Actual behavior

RuboCop triggers `Style/FormatStringToken` error:

```
spec/classes/sshd_spec.rb:3:55: C: Style/FormatStringToken: Prefer annotated tokens (like %<foo>s) over unannotated tokens (like %s).
    .with_content(%r{^    ChrootDirectory /var/chroot/%u$})
                                                      ^^
```

## Steps to reproduce the problem

The snippet above triggers the described offense.

## RuboCop version

```
$ rubocop -V
0.52.0
```

**Comment by jeremywadsack** [comment]
Another instance of this is the use of `printf` style format strings which I think are pretty common — in my experience much more common than format strings (outside of Rails I18n).

```
Style/FormatStringToken: Prefer annotated tokens (like %<foo>s) over unannotated tokens (like %s).
      @partition ||= dataset.table("#{table.table_id}$#{@date.strftime('%Y%m%d')}")
```

[Tagging #4539 for x-ref.]

**Comment by envygeeks** [comment]
We ran into this with `strftime` too.

**Comment by pocke** [comment]
@pegasd Thanks for your bug report! I'll fix it.

@jeremywadsack @envygeeks 
The `strftime`'s issue will be fixed by #5230 
See https://github.com/bbatsov/rubocop/issues/5223 https://github.com/bbatsov/rubocop/pull/5230

**Comment by davidhrbac** [comment]
We ran into this with select like
```
 "date_format(IF(dateEfficiency>NOW(),dateEfficiency,dateExpiration),'%Y%m%d%H%i%s')" \
                                                                              ^^
```

**Comment by pegasd** [comment]
@davidhrbac Yeah, I think that particular case is discussed in #5230

**Comment by ctm** [comment]
`scanf` also takes a string that uses formatting characters:
```
irb(main):001:0> s = '12'
=> "12"
irb(main):002:0> require 'scanf'
=> true
irb(main):003:0> s.scanf('%f').first
=> 12.0
irb(main):004:0> s.scanf('%<ignored>f').first
=> nil
```

**Comment by jeremywadsack** [comment]
@pocke Thanks for linking that fix for `strftime` — sorry I missed that when looking for previous reports.

**Comment by envygeeks** [comment]
Thanks for the quick fix!

**Comment by agross** [comment]
I see this as a false positive for `"foo/bar.rb".pathmap('%d/baz/%f')`.

**Comment by agross** [comment]
Also in this spec:
```rb
context 'something' do
  let(:content) { template % address }
                           ^^
  let(:expected) { template % something(address) }

  let(:address) { 'example@example.com' }

  context 'text' do
    let(:template) { 'prefix %s suffix' }

    it 'yields encoded content' do
      expect(subject.run(content)).to eq(expected)
    end
  end
end
```

**Comment by mikegee** [comment]
@lionel218 the cop wants you to use something like:

```ruby
format('%<day>s %<start>s-%<end>s', day: starts_at.strftime('%a'), start: starts_at.strftime('%l%P'), end: ends_at.strftime('%l%P').strip)
```

**Comment by lionel218** [comment]
Hi @mikegee ,  Just solved this issue.

Thanks for your kindness.

**Comment by lionel218** [comment]
Hi @mikegee , How can I solve this issue?
 Style/FormatStringToken: Prefer annotated tokens (like %<foo>s) over unannotated tokens (like %s).
 starts_at.strftime('%A')

**Comment by lionel218** [comment]
Please help me for this issue.

**Comment by mikegee** [comment]
@lionel218 I failed to duplicate the issue with the example you provided. Please open a new issue with all the details if you think this bug is still in Rubocop. Thanks!

**Comment by lionel218** [comment]
@mikegee , Thanks for your kindness.
But I can't create new issue because the submit issue button is disabled now.

Please check for my code and rubocop issue.

def event_date(open_house)
    format('%<day>s %<start>s-%<end>s', day: open_house.starts_at.strftime('%a'),
                                        start: open_house.starts_at.strftime('%l'),
                                        end: open_house.ends_at.strftime('%l %p').strip)
end
This is my helper action, but just got this issue on rubocop-0.53.0, ruby-2.4.2, rails-5.1.1

app/helpers/open_houses_helper.rb:3:77: C: Style/FormatStringToken: Prefer annotated tokens (like %<foo>s) over unannotated tokens (like %s).
    format('%<day>s %<start>s-%<end>s', day: open_house.starts_at.strftime('%a'),
                                                                            ^^
app/helpers/open_houses_helper.rb:5:78: C: Style/FormatStringToken: Prefer annotated tokens (like %<foo>s) over unannotated tokens (like %s).
                                        end: open_house.ends_at.strftime('%l %p').strip)

**Comment by mikegee** [comment]
@lionel218 I confirmed with that code and opened another issue for you: https://github.com/bbatsov/rubocop/issues/5630

**Comment by lionel218** [comment]
Thanks so much @mikegee

**Comment by Seenox** [comment]
Sorry for bumping an old issue, but since it shows up in first page of Google if you search for:
```Prefer annotated tokens (like `%<foo>s`) over template tokens (like `%{foo}`)```
I thought I'd share what has helped me to work my way around that.

My code:
```ruby
[...]
    when /RedHat|CentOS|Fedora|SLES/                                                                                  
      installed_packages = Facter::Core::Execution.exec('rpm -qa --qf "%{NAME} %{VERSION}\n"').split('\n')
    when /Debian|Ubuntu/
      ...
[...]
```
Was throwing an error:
```rubocop: Prefer annotated tokens (like `%<foo>s`) over template tokens (like `%{foo}`).```

Disabling `Style/FormatStringToken` right before and re-enabling it right after have helped my pipeline to succeed:
```ruby
[...]
    when /RedHat|CentOS|Fedora|SLES/                                                                                  
      # rubocop:disable Style/FormatStringToken
      installed_packages = Facter::Core::Execution.exec('rpm -qa --qf "%{NAME} %{VERSION}\n"').split('\n')
      # rubocop:enable Style/FormatStringToken
    when /Debian|Ubuntu/
[...]
```
It may not be the correct/best way to solve it, but it was the quickest for someone like me who has little knowledge about `robocup`.
